### PR TITLE
Added TextTree parser

### PR DESCRIPTION
**Pull request by dbarrosop**
First attempt to adding a parser that is able to create a tree like structure from unstructured text. What I expect with this is to be able to do things like:

```
fa1_desc = parsed["interface"]["fa1"]["description"]["#text"]
for ip, data in  parsed["interface"]["fa1"]["ip"]["address"].items():
    print("ip: {}, mask: {}".format(ip, data["#text"])
```

It also has `#standalone` key where I can signal that it's the end of the command.

The idea is to do this:

```
interface Fa1
   switchport
   switchport mode access
   switchport access vlan 2,3,4,5

interface
  Fa1:
    #standalone: true
    switchport:
      #standalone: true
      mode:
        #text: access
        access:
          #standalone: true
      access:
       #text: vlan 2,3,4,5
       vlan:
          #standalone: true
          #text: 2,3,4,5
```

So thanks to `#standalone` there is an easy way to see that `switchport` is an independent command and not just built by `switchport mode access`.

There is also a `#list` where the order is preserved. For example:

```
ip access-list extended SOMETHING
   deny tcp any any eq 80
   deny tcp any any eq 443

ip:
  access-list:
    extended:
      SOMETHING:
        #list:
          - deny:
              #text: "tcp any any eq 80", 
              tcp:
                #text: "any any eq 80", 
                any:
                ...
          - deny:
              #text: "tcp any any eq 443", 
              tcp:
                #text: "any any eq 443", 
                any:
                ...
```

So you should be able to do both `config["ip"]["access-list"]["extended"]["SOMETHING"]["deny"]...` or `config["ip"]["access-list"]["extended"]["SOMETHING"][0]["deny"]`.

You can see how a full running config maps to a dict here: https://github.com/napalm-automation/napalm-yang/tree/6958d8ac9235219006c49e9702efe192105761b1/test/unit/text_tree/test_case_1

I think this is getting there. The original parser for interfaces was extremely more complex than it is now with this parser: https://github.com/napalm-automation/napalm-yang/pull/76/commits/6958d8ac9235219006c49e9702efe192105761b1

### CODE REVIEW

**Comment by dbarrosop** [comment]
@matejv following on the discussion we had on #75 I have implemented the TextTree parser and migrated EOS profiles. Would you mind reviewing? I think the new parser is simpler to use. I also think it should be less error prone and we shouldn't see issues like the one you found.

In addition, I got rid of `flatten` and `expand_list` (used on junos) and replaced those with an expansion mechanism you can use in the path:

https://github.com/napalm-automation/napalm-yang/pull/76/files#diff-9b3975756ac2a43b7d34d5091a73a95e

**Comment by matejv** [comment]
That looks much nicer at first glance! Still going through everything. 

I did notice one something I think is a bug - all elements under `#list` with the same key have the same value:

```python
txt='''TestAcl
 permit 1.1.1.1
 deny 2.2.2.2
 permit 2.2.2.2
'''
config = parse_indented_config(txt.splitlines())
config['TestAcl']['#list']
# returns:
[{'permit': OrderedDict([('#text', '2.2.2.2'),
               ('1.1.1.1', OrderedDict([('#standalone', True)])),
               ('2.2.2.2', OrderedDict([('#standalone', True)]))])},
 {'deny': OrderedDict([('#text', '2.2.2.2'),
               ('2.2.2.2', OrderedDict([('#standalone', True)]))])},
 {'permit': OrderedDict([('#text', '2.2.2.2'),
               ('1.1.1.1', OrderedDict([('#standalone', True)])),
               ('2.2.2.2', OrderedDict([('#standalone', True)]))])}]

config['TestAcl']['#list'][0] == config['TestAcl']['#list'][2]
# first and last element are equal:
True
```

Probably need to re-init some dict at the start of some loop?

Looking through the rest now.

**Comment by dbarrosop** [comment]
Yes, I think it's the same error I encountered when implementing the parser for the openconfig-vlan model. This should fix it:

https://github.com/napalm-automation/napalm-yang/pull/77/commits/83554bb78f5d144f4e18d10f298151ceeef4ff23

**Comment by dbarrosop** [comment]
Oh, no. It's not the same issue. Looking into it. Good catch :)

**Comment by matejv** [comment]
Ugh. I can't figure out how to have a key from two consecutive words of configuration.

IOS native (yeah I know, sorry :)
```
ip route 7.7.7.7 255.255.255.255 1.1.1.1
```

Parsed dict after `parse_indented_config` runs (in path `resp[0]['ip']['route']`):
```
{
    "7.7.7.7": {
        "#text": "255.255.255.255 1.1.1.1", 
        "255.255.255.255": {
            "#text": "1.1.1.1", 
            "1.1.1.1": {
                "#standalone": true
            }
        }
    }, 
}
```

Following your EOS example I can set the key to the `7.7.7.7` part, but don't know how to add the netmask after it:
```yaml
static:
    _process:
        - path: ""
          regexp: "(?P<value>^\\d+\\.\\d+\\.\\d+\\.\\d+)"
          when: "{{ network_instance_key == 'global' }}"
```

Sorry, I'm kind of lost in the code now, can't even tell if this is possible or not.

**Comment by dbarrosop** [comment]
Not possible now but shouldn't be hard to fix. For example, the rule might look like:

```
- path: ip.route.?prefix.?mask
   key: "{{prefix}}/{{mask}}"
```

The way that path works right now should make it return something like:

```
- prefix: 7.7.7.7
   mask: 255.255.255.255
   other_stuff: whatevs
- prefix: 8.8.8.0
   mask: 255.255.255.0
   other_stuff: whatevs
```

So we only need to add support to combine fields in the key. I think we can reuse what your `post_processing_filter` code does for something like this. If you have the parsing/translation working with the textparser and you also have working tests, please, send a PR against this branch and I will merge and fix them so it works with the TextTree parser.

**Comment by matejv** [comment]
Ah, I see. I suspected the question mark path syntax would be of use here, I was just missing how to combine multiple fields. So it's not just me being dense then :)

I have mappings for ios network-instance mostly done with the same features as eos has (plus IPv6 support and a few minor additions we use extensively) [here](https://github.com/ArnesSI/napalm-yang/tree/ios-routing). I'm just missing a few BGP attributes to achieve feature parity with EOS. I'll rebase against this branch and open a PR.

**Comment by dbarrosop** [comment]
Yeah, I think the new path behavior will be quite handy :P I just can use the `openconfig-interfaces` model on IOS to work on this as the `ip address` command will have the same issue so no rush, I will use that one to work on it :)

File: napalm_yang/mappings/ios/parsers/config/openconfig-if-ip/ipv4.yaml [hunk 1]
```diff
@@ -1,34 +1,35 @@
...
+                  path: "switchport"
         mtu:
             _process: not_implemented
     addresses:
         _process: unnecessary
         address:
             _process:
-                - regexp: "(?P<block>ip address (?P<key>(?P<ip>.*)) (?P<prefix>([0-255]|\\.)*)(?P<secondary> secondary)*)$"
+                - path: "ip.address.?prefix.?mask"
+                  regexp: "^(?P<value>\\d+\\.\\d+\\.\\d+\\.\\d+)"
```
Comment (dbarrosop) [inline napalm_yang/mappings/ios/parsers/config/openconfig-if-ip/ipv4.yaml]: The regexp helps us filtering things like `ip address dhcp`.

**Comment by dbarrosop** [comment]
@matejv haven't fixed the composed key yet but you can see here how to leverage the `?` to flatten dictionaries and extract data:

https://github.com/napalm-automation/napalm-yang/pull/76/files#diff-1af988e1418bcef8435d1994e7c8e65dR15

Will check next how to do that.

File: napalm_yang/mappings/ios/parsers/config/openconfig-if-ip/ipv4.yaml (hunk 1 above)
Comment (matejv) [inline napalm_yang/mappings/ios/parsers/config/openconfig-if-ip/ipv4.yaml]: But what value is the regex applied against? `prefix` or `mask` part of the path?

Comment (matejv) [inline napalm_yang/mappings/ios/parsers/config/openconfig-if-ip/ipv4.yaml]: Ah, I think I got it. So with a question mark you basically assign whatever key is below `address` to a field name `prefix`.  But that key must match `regex`, plus `prefix` will be set to the `value` named capture group? And now you can reference that value by asking for `prefix`. And the same goes for `mask` as well. Correct?

But what will happen if I have `ip address 1.1.1.1 dog`?

**Comment by dbarrosop** [comment]
@matejv check ab90eb44d179d30f42a5a6d0696e638d1a0671ab. I will revert it and implement it properly but the idea is to replace all different ways we have to process the keys now ("key", "path", "key_value", "composite_key" and "post_process_filter") with just that one and single mechanism. What do you think?

**Comment by matejv** [comment]
Well I wanted `post_process_filter`  to be able to access `extra_vars` in TextParser in the same leaf as they were being matched. As far as I can see, in TextTree this is no longer an issue.

I'm all for reducing the number of possible parameters a mapping can have :)

My concern is performance - if every key needs to be set by rendering a Jinja template this can take a while if parsing a large configuration. Especially since I think most key are just a simple value.

**Comment by dbarrosop** [comment]
> My concern is performance - if every key needs to be set by rendering a Jinja template this can take a while if parsing a large configuration. Especially since I think most key are just a simple value.

That's fine, right now we should focus on a stable/sane/easy_to_use API. We should profile the code and apply optimizations later on. In this particular case there are a few optimizations we could do, for example, we could avoid using jinja2 `if "{{" not in jinja_key`.

**Comment by dbarrosop** [comment]
@matejv check: de0770f. That should be a better implementation and feels quite generic so it should work for any use case. It also uses the jinja2 framework when necessary so it shouldn't have an unnecessary impact on performance.

**Comment by matejv** [comment]
Sorry other stuff kept coming up today, didn't manage to do anything more than browse through the new changes. This is looking great at first glance. Tomorrow I'll try to port my IOS parser to TextTree parser and see if I find anything that's missing or unclear.

**Comment by dbarrosop** [comment]
No worries, I will start updating in parallel the documentation to reflect how things work now. I think right now the DSL for the parser is quite clean and ready but I guess documenting everything will help identifying redundancies and complexities. Note in the last few commits that now `value` (leaf) behaves in the same way as `key` (lists) so there is no need for post_processors.

I also profiled it and although there are probably some things we can improve this new behavior didn't introduce any noticeable overhead.

**Pull request merged**